src/odemis/driver/actuator.py

```diff
@@ -2566,82 +2568,83 @@ def __init__(self, name, role, parent, dependencies, rng, executor, **kwargs):
 
-        # Set the initial focus active position with range 0
+        # Set the initial focus active position with range min
         self._metadata[model.MD_FAV_POS_ACTIVE] = {'z': self._range[0]}
-        logging.info("%s changed to %s", model.MD_FAV_POS_ACTIVE, self._metadata[model.MD_FAV_POS_ACTIVE])
+        logging.info("Focus FAV_POS_ACTIVE changed to %s.", self._metadata[model.MD_FAV_POS_ACTIVE])
 
         self.referenced = model.VigilantAttribute({}, readonly=True)
         # listen to reference changes from the underlying stage
-        logging.debug("Subscribing to reference of dependency %s", self._lensz.name)
-        self._lensz.referenced.subscribe(self._ondepStageReferenced)
-        self._ondepStageReferenced(self._lensz.referenced.value)
+        self._lensz.referenced.subscribe(self._onLensReferenced, init=True)
 
-    def _ondepStageReferenced(self, cref):
+    def _onLensReferenced(self, cref):
         """
-        Update the referenced value from the underlying stage
+        Update the referenced value from the underlying lens stage
         """
         # Directly set the value from the updated argument
         if 'z' in cref:
-            self.referenced._set_value(cref['z'], force_write=True)
+            self.referenced._set_value({'z': cref['z']}, force_write=True)
 
     def updateMetadata(self, md):
+        # Prevent manual update of focus FAV_POS_ACTIVE and FAV_POS_DEACTIVE
         if model.MD_FAV_POS_ACTIVE in md or model.MD_FAV_POS_DEACTIVE in md:
-            raise ValueError("%s and %s cannot be set manually" % model.MD_FAV_POS_ACTIVE % model.MD_FAV_POS_DEACTIVE)
+            raise ValueError("Focus FAV_POS_ACTIVE and FAV_POS_DEACTIVE cannot be set manually.")
 
         super(LinkedHeightFocus, self).updateMetadata(md)
-        # self._metadata.update(md)
+        # Re-update focus position if POS_COR is modified
         if model.MD_POS_COR in md:
             self._updatePosition(self._lensz.position.value)
 
     def _updatePosition(self, lens_pos):
         """
-        update the position VA from the underlying dependency
-        (the method is public to be used from the parent)
+        Update the position VA from the underlying dependency
+        :param lens_pos: (dict string-> float) the lens Z value
         """
         # Only update position when MD_POS_COR is already configured
         if model.MD_POS_COR not in self._metadata:
-            logging.error("%s not in metadata" % model.MD_POS_COR)
+            logging.error("Focus POS_COR is not found in metadata.")
             return
-        focus_val = self._getFocusValue(target_lensz=lens_pos['z'])
-        logging.debug("Reporting focus position %s", focus_val)
+        if util.almost_equal(lens_pos['z'], self._lensz.getMetadata()[model.MD_FAV_POS_DEACTIVE]['z'], atol=1e-3):
+            # Set focus FAV_POS_DEACTIVE when the lens pos is FAV_POS_DEACTIVE
```

Comment:
Had to add this special check to set focus `FAV_POS_DEACTIVE` when the lens pos is `FAV_POS_DEACTIVE` (as the reverse is done on when move is initialed) so it's wrong value to get it from `_getFocusValue`